File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
from collections import deque
def solution(arr):
    direction = [(0,1),(1,0),(0,-1),(-1,0)]
    ans = []
    n = len(arr)
    m = len(arr[0])
    visit = [[False]*m for _ in range(n)]
    def bfs(x,y):
        nonlocal visit
        q = deque()
        visit[x][y] = True
        q.append((x,y))
        s = int(arr[x][y])
        while q:
            cx,cy = q.popleft()
            for i in range(4):
                nx,ny = cx+direction[i][0],cy+direction[i][1]
                if 0<=nx<n and 0<=ny<m and not visit[nx][ny] and arr[nx][ny] != 'X':
                    visit[nx][ny] = True
                    s += int(arr[nx][ny])
                    q.append((nx,ny))
        print(s)
        return s
        
    for i in range(n):
        for j in range(m):
            if not visit[i][j] and arr[i][j] != 'X':
                ans.append(bfs(i,j))
    return [-1] if len(ans) == 0 else sorted(ans)

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.java
import java.util.*;
class Solution {
    int[][] dir = {{0,1},{0,-1},{1,0},{-1,0}};
    int m;
    int n;
    boolean[][] visit;
    char[][] arr;
    List<Integer> ans = new ArrayList<>();
    public int[] solution(String[] maps) {
        n = maps.length;
        m = maps[0].length();
        visit = new boolean[n][m];
        arr = new char[n][m];
        // arr 초기화
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                arr[i][j] = maps[i].charAt(j);
            }
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (!visit[i][j] && arr[i][j] != 'X') bfs(i,j);
            }
        }
        if (ans.size() == 0) return new int[] {-1};
        Collections.sort(ans);
        int[] answer = new int[ans.size()];
        for (int i = 0; i < ans.size(); i++) answer[i] = ans.get(i);
        return answer;
    }
    
    public void bfs(int x, int y) {
        Queue<int[]> q = new ArrayDeque<>();
        visit[x][y] = true;
        int cnt = Integer.valueOf(arr[x][y]) - '0';
        q.add(new int[] {x,y});
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int cx = cur[0];
            int cy = cur[1];
            for (int[] d : dir) {
                int nx = cx + d[0];
                int ny = cy + d[1];
                if (0 <= nx && nx < n && 0 <= ny && ny < m) {
                    if (!visit[nx][ny] && arr[nx][ny] != 'X') {
                        cnt += Integer.valueOf(arr[nx][ny]) - '0';
                        visit[nx][ny] = true;
                        q.add(new int[] {nx,ny});
                    }
                }
            }
        }
        ans.add(cnt);
    }
}